HPCC-14000 Better fix for incorrect balanced splitters

ghalliday opened the pull request:
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

ghalliday committed:
HPCC-14000 Better fix for incorrect balanced splitters

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday commented:
@richardkchapman I think it might be worth talking through this problem to see if anyone has any other bright ideas when we next meet.  It will be easier to discuss in person.
@jakesmith @jamienoss please review.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jamienoss commented:
@ghalliday I guess we forgot to look at this at the offsite...

Michael-Gardner committed:
HPCC-14303 Added TOP_LEVEL_PROJECT variable to CMakeLists.txt

Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

Commit:
HPCC-14292 Auto refresh button needs tooltip

Add a tooltip to auto refresh and whether it is on/off.

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexis.com>

ghalliday commented:
@richardkchapman please review

richardkchapman commented on ecl/hqlcpp/hqlresource.cpp:
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
+void EclResourcer::removeDuplicateIndependentLinks(CSplitterInfo & connections, ResourcerInfo & info)
+{
+    IHqlExpression * expr = info.original;
+    bool again = true;
```
This feels like an odd way of coding a do-wihle loop...

richardkchapman commented on ecl/hqlcpp/hqlresource.cpp:
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
+
+                //Removing duplicate links has turned the source item into a pass-through
+                //replace references to the sink activity with references to its sink
+                //to possible allow more to be removed
```
Grammar - possible -> possibly ?

richardkchapman commented on ecl/hqlcpp/hqlresource.cpp:
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
+                }
+
+                //Removing duplicate links has turned the source item into a pass-through
+                //replace references to the sink activity with references to its sink
```
Capital R on replace would make it slightly easier to read

richardkchapman commented on ecl/hqlcpp/hqlresource.cpp:
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
+            CSplitterLink & cur = info.balancedLinks.item(i);
+            if (cur.hasSource(expr))
+            {
+                IHqlExpression * sink = cur.queryOther(expr);
```
No chance of NULL returns here?

ghalliday replied:
No, but I'll add a check.

richardkchapman commented:
@ghalliday Some minor questions/comments

mckellyln committed:
HPCC-14564 Disable -std=c++11 for eclrtl.cpp when on debian and boost < 1.48

Signed-off-by: Mark Kelly <mark.kelly@lexisnexis.com>

richardkchapman committed:
HPCC-14576 Fix incorrect code detected by compiler warning

Code was incorrect, and would have resulted in libraries that started 'l' but
did not start 'lib' not having the 'lib' added when it should have. Unlikely
to have caused a problem in practice.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14578 Fix jlib warnings

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14577 OSX builds warn about missing CLOCK_MONOTONIC

Also improve accuracy of timers on OSX.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

jakesmith committed:
HPCC-14584 Avoid potential exception in call from destructor

With the c++11 semantics enabled, throwing an exception directly
or indirectly from a desctructor that hasn't been declared with
noexcept(false) will terminate the program.

The locking code prior to this change had a wrapper class that
during destruction called a method that could throw an exception.
Avoid the use of this simple wrapper class and call the methods
it was calling in the ctor and dtor directly to avoid being hit
by the new c++11 semantics.
NB: alternative fix was to add noexcept(false) to dtor, but
avoiding the possibility of an exception in destructor altogether
is safer.

Signed-off-by: Jake Smith <jake.smith@lexisnexis.com>

ghalliday committed:
Merge pull request #7983 from richardkchapman/register-warning2

HPCC-14578 Fix jlib warnings

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #7982 from richardkchapman/monotonic-osx2

HPCC-14577 OSX builds warn about missing CLOCK_MONOTONIC

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #7979 from richardkchapman/strncmp-warning

HPCC-14576 Fix incorrect code detected by compiler warning

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
HPCC-14586 Support HINT on COUNT() and DATASET([])

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Merge pull request #7978 from mckellyln/hpcc-14564

HPCC-14564 Disable -std=c++11 for eclrtl.cpp with g++ and clang due to boost (< 1.48) compile issue

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #7988 from jakesmith/hpcc-14584

HPCC-14584 Avoid potential exception in call from destructor

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

g-pan committed:
HPCC-14309 DOCS:Keywords Global Search in ECL Watch.

FIX HPCC-14309 DOCS:Keywords Global Search in ECL Watch
Document keywords in ECL watch global search

Signed-off-by: G Panagiotatos <greg.panagiotatos@lexisnexis.com>

ghalliday committed:
Merge pull request #7915 from Michael-Gardner/HPCC-14303

HPCC-14303 Added TOP_LEVEL_PROJECT variable to CMakeLists.txt

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
HPCC-14602 Remove unused legacy stats tag code

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

jamienoss committed:
HPCC-14605 Remove bison output files before rebuilding

Signed-off-by: James Noss <james.noss@lexisnexis.com>

g-pan committed:
HPCC-14309 DOCS:Keywords Global Search in ECL Watch.

FIX HPCC-14309 DOCS:Keywords Global Search in ECL Watch
Document keywords in ECL watch global search

Signed-off-by: G Panagiotatos <greg.panagiotatos@lexisnexis.com>

jamienoss committed:
HPCC-14612 ECLlocation's constructor should initialize its members

Signed-off-by: James Noss <james.noss@lexisnexis.com>

ghalliday committed:
HPCC-14613 Fix trivial windows build errors

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Merge pull request #7998 from ghalliday/issue14613

HPCC-14613 Fix trivial windows build errors

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #7997 from jamienoss/issue14612-ECLlocation-constructor-init-members

HPCC-14612 ECLlocation's constructor should initialize its members

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #7995 from jamienoss/issue14605-remove-bison-file-on-rebuild

HPCC-14605 Remove bison output files before rebuilding

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

richardkchapman committed:
Merge pull request #7996 from g-pan/H14309-ECLWaSearch

HPCC-14309 DOCS:Keywords Global Search in ECL Watch

Reviewed-By: Jim DeFabia <james.defabia@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #7991 from ghalliday/issue14586

HPCC-14586 Support HINT on COUNT() and DATASET([])

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #7992 from ghalliday/issue14602

HPCC-14602 Remove unused legacy stats tag code

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #7954 from miguelvazq/HPCC-14292

HPCC-14292 Auto refresh button needs tooltip

Reviewed-By: Gordon Smith <gordon.smith@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

ghalliday commented:
See #8001 for new version
